File: sdks/go/identity.go
package croniq

import (
	"crypto/rand"
	"encoding/hex"
	"errors"
	"fmt"
	"log/slog"
	"os"
	"path/filepath"
	"strings"
)

// Resolution order for [ResolveRunnerID].
const (
	defaultDataDir = "/var/lib/croniq-runner"
	stateFileName  = "runner-id"
)

// ResolveRunnerID returns a stable runner_id across container recreates.
//
// Resolution order:
//
//  1. RUNNER_ID env var — explicit operator override, used as-is.
//  2. State file at ${CRONIQ_RUNNER_DATA_DIR:-/var/lib/croniq-runner}/runner-id
//     — read if it exists and non-empty.
//  3. Generate "{prefix}-{8-hex}" and persist to the state file.
//
// If persistence fails (no volume mounted, read-only filesystem, …),
// falls back to a volatile hostname-derived id and logs a warning. The
// runner still starts; only the cross-recreate stability is lost.
func ResolveRunnerID(prefix string) string {
	if v := strings.TrimSpace(os.Getenv("RUNNER_ID")); v != "" {
		return v
	}
	dir := strings.TrimSpace(os.Getenv("CRONIQ_RUNNER_DATA_DIR"))
	if dir == "" {
		dir = defaultDataDir
	}
	return resolveOrPersist(prefix, dir)
}

func resolveOrPersist(prefix, dataDir string) string {
	statePath := filepath.Join(dataDir, stateFileName)

	if buf, err := os.ReadFile(statePath); err == nil {
		id := strings.TrimSpace(string(buf))
		if id != "" {
			return id
		}
	}

	id := generateID(prefix)
	if err := persistID(statePath, id); err != nil {
		fallback := volatileFallback(prefix)
		slog.Warn(
			"could not persist runner identity — falling back to volatile ID. Mount a writable volume at this path to make runner_id stable across container recreates.",
			"path", statePath,
			"error", err,
			"runner_id", fallback,
		)
		return fallback
	}
	slog.Info("generated new runner identity and persisted to state file",
		"path", statePath,
		"runner_id", id,
	)
	return id
}

func generateID(prefix string) string {
	var buf [4]byte
	if _, err := rand.Read(buf[:]); err != nil {
		// crypto/rand failure is "the OS is broken" territory — fall
		// back to something rather than panicking, the runner can
		// still start.
		return fmt.Sprintf("%s-%08x", prefix, os.Getpid())
	}
	return fmt.Sprintf("%s-%s", prefix, hex.EncodeToString(buf[:]))
}

func persistID(path, id string) error {
	if path == "" {
		return errors.New("empty state-file path")
	}
	if dir := filepath.Dir(path); dir != "" {
		if err := os.MkdirAll(dir, 0o755); err != nil {
			return err
		}
	}
	return os.WriteFile(path, []byte(id), 0o644)
}

func volatileFallback(prefix string) string {
	if host := strings.TrimSpace(os.Getenv("HOSTNAME")); host != "" {
		return prefix + "-" + host
	}
	return generateID(prefix)
}

File: sdks/typescript/src/identity.ts
import { randomBytes } from 'node:crypto';
import { existsSync, mkdirSync, readFileSync, writeFileSync } from 'node:fs';
import { homedir, hostname } from 'node:os';
import { join } from 'node:path';

import type { Logger } from './logger.js';

export interface IdentityResolverInput {
  /** Explicit runner-id from configuration. */
  runnerId?: string | undefined;
  /** Prefix used when generating a fresh id. */
  runnerIdPrefix: string;
  /** Optional data dir override. Falls back to env var, then platform default. */
  runnerDataDir?: string | undefined;
}

/**
 * Resolves the runner ID with the same precedence as the .NET / Rust SDKs:
 *
 *  1. Explicit option
 *  2. `RUNNER_ID` env var
 *  3. Persisted `runner-id` state file under {@link IdentityResolverInput.runnerDataDir}
 *     (or `CRONIQ_RUNNER_DATA_DIR`, or the platform default)
 *  4. Newly generated `{prefix}-{hex8}`, persisted to the state file
 *
 * If persistence fails (filesystem read-only etc.), returns a deterministic
 * `{prefix}-{hostname}` fallback so the runner can still start.
 */
export function resolveRunnerId(input: IdentityResolverInput, logger: Logger): string {
  if (input.runnerId && input.runnerId.length > 0) return input.runnerId;

  const fromEnv = process.env.RUNNER_ID;
  if (fromEnv && fromEnv.length > 0) return fromEnv;

  const dataDir = resolveDataDir(input.runnerDataDir);
  const idFile = join(dataDir, 'runner-id');

  try {
    if (existsSync(idFile)) {
      const persisted = readFileSync(idFile, 'utf8').trim();
      if (persisted.length > 0) return persisted;
    }
  } catch (err) {
    logger.warn(`could not read persisted runner ID from ${idFile}`, { error: String(err) });
  }

  const generated = `${input.runnerIdPrefix}-${randomBytes(4).toString('hex')}`;
  try {
    mkdirSync(dataDir, { recursive: true });
    writeFileSync(idFile, generated, 'utf8');
  } catch (err) {
    logger.warn(`could not persist generated runner ID to ${idFile}`, { error: String(err) });
    return `${input.runnerIdPrefix}-${hostname().toLowerCase()}`;
  }

  return generated;
}

function resolveDataDir(override?: string): string {
  if (override && override.length > 0) return override;
  const fromEnv = process.env.CRONIQ_RUNNER_DATA_DIR;
  if (fromEnv && fromEnv.length > 0) return fromEnv;

  // Platform-appropriate state directory.
  // On Linux/macOS use $XDG_STATE_HOME or ~/.local/state; on Windows use %LOCALAPPDATA%.
  if (process.platform === 'win32') {
    const localAppData = process.env.LOCALAPPDATA ?? join(homedir(), 'AppData', 'Local');
    return join(localAppData, 'croniq-runner');
  }
  const xdgStateHome = process.env.XDG_STATE_HOME;
  const base = xdgStateHome && xdgStateHome.length > 0
    ? xdgStateHome
    : join(homedir(), '.local', 'state');
  return join(base, 'croniq-runner');
}
